src/softmax.jl

```diff
@@ -55,33 +73,32 @@ function softmax!(out::AbstractArray{T}, x::AbstractArray; dims = 1) where {T}
     else
         @. out = ifelse(isequal(max_,Inf), ifelse(isequal(x,Inf), 1, 0), exp(x - max_))
     end
-    out ./= sum(out; dims = dims)  # could re-use max_ when dims != (:) and eltype(x) == T.
+    y = out ./= sum(out; dims = dims)
 end
 
-∇softmax(Δ::AbstractArray{T}, x::AbstractArray, y::AbstractArray{S}; dims = 1) where {T,S} = 
-    ∇softmax!(similar(y, promote_type(T, S)), Δ, x, y; dims = dims)
-∇softmax(Δ, x, y; dims = 1) = ∇softmax(unthunk(Δ), x, y, dims = dims)
-
-# Can introduce at the end of deprecation cycle of ∇softmax!(out, Δ, x; dims = 1)
-# ∇softmax!(Δ, x, y; dims = 1) = ∇softmax!(Δ, Δ, x, y; dims = dims)
-
-function ∇softmax!(out::AbstractArray, Δ::AbstractArray, 
-                    x::AbstractArray, y::AbstractArray; dims = 1)
-    out .= Δ .* y
-    out .= out .- y .* sum(out; dims = dims)
+function ∇softmax(dy::AbstractArray{T}, x::AbstractArray, y::AbstractArray{S}; dims = 1) where {T,S}
```

Comment:
The only reason this accepts `x` is to disambiguate from previous signature `∇softmax(Δ, x; dims = 1)`. Maybe we should allow you to pass anything (or `nothing`) for `x`, so as not to close over it.
```suggestion
function ∇softmax(dy::AbstractArray{T}, x, y::AbstractArray{S}; dims = 1) where {T,S}
```
Or maybe we should just give this a new name & a sensible signature?

```suggestion
function softmax∇x(dy::AbstractArray{T}, y::AbstractArray{S}; dims = 1) where {T,S}
```

Reply:
I agree with the change but since we export `∇softmax` that would be breaking.

Reply:
ok I saw you put a deprecation in place, fine

Reply:
maybe we should remove the exports of gradients at some point since we define the rrules. I don't know if there are any consumers of those functions (e.g. Nabla.jl, ...).

Reply:
Oh I didn't notice it was exported.

I think we should probably chase down most users before merging. Tracker still uses the 2-arg ∇softmax...

Reply:
If the question is about naming schemes, I like what we have now with `∇conv_filter` and `∇conv_data`. That would be a tad more verbose than `softmax∇x`, but it would save some guesswork on what the positioning of each nabla means.

Reply:
I just thought that `∇softmax_x` seemed a bit ugly, maybe `∇softmax_data` is better... the main goal was just not to re-use the existing name.